Sources/Jobs/Scheduler/RecurrenceRule/RecurrenceRule.swift

```diff
@@ -0,0 +1,413 @@
+            }
+
+        }
+
+        if ruleEvaluationState == .passing {
+            return (isValid: true, ruleTimeUnitFailedOn)
+        } else {
+            return (isValid: false, ruleTimeUnitFailedOn)
+        }
+    }
+
+    private func isYearConstraintPossible(date: Date) -> Bool {
+        let currentYear = date.year()!
+
+        if yearConstraint.isConstraintActive {
+            for year in yearConstraint.setConstraint {
+                if year >= currentYear {
+                    return true
+                }
+            }
+
+            if let rangeConstraint = yearConstraint.rangeConstraint {
+                if rangeConstraint.contains(currentYear) {
+                    return true
+                }
+            }
+
+            if yearConstraint.stepConstraint != nil {
+                return true
+            }
+
+            return false
+        } else {
+            return true
+        }
+    }
+
+    private func resolveTimeUnitOfActiveConstraintWithLowestCadenceLevel() -> RecurrenceRuleTimeUnit? {
+        var activeConstraintTimeUnitWithLowestCadenceLevel: RecurrenceRuleTimeUnit? = nil
+
+        for ruleTimeUnit in recurrenceRuleTimeUnits {
+            let constraint = resolveConstraint(ruleTimeUnit)
+            if constraint.isConstraintActive {
+                activeConstraintTimeUnitWithLowestCadenceLevel = ruleTimeUnit
+            }
+        }
+
+        return activeConstraintTimeUnitWithLowestCadenceLevel
+    }
+
+    private func resolveNextValidValue(for ruleTimeUnit: RecurrenceRuleTimeUnit, date: Date) throws -> Int {
+        guard let currentValue = date.resolveDateComponentValue(for: ruleTimeUnit) else {
+            throw RecurrenceRuleError.couldNotResolveDateComponentValueFromRecurrenceRuleTimeUnit
+        }
+
+        switch ruleTimeUnit {
+        case .second:
+            return secondConstraint.nextValidValue(currentValue: currentValue)!
+        case .minute:
+            return minuteConstraint.nextValidValue(currentValue: currentValue)!
+        case .hour:
+           return hourConstraint.nextValidValue(currentValue: currentValue)!
+        case .dayOfWeek:
+            return dayOfWeekConstraint.nextValidValue(currentValue: currentValue)!
+        case .dayOfMonth:
+            return dayOfMonthConstraint.nextValidValue(currentValue: currentValue)!
+        case .weekOfMonth:
+            return weekOfMonthConstraint.nextValidValue(currentValue: currentValue)!
+        case .weekOfYear:
+            return weekOfYearConstraint.nextValidValue(currentValue: currentValue)!
+        case .month:
+            return monthConstraint.nextValidValue(currentValue: currentValue)!
+        case .quarter:
+            return quarterConstraint.nextValidValue(currentValue: currentValue)!
+        case .year:
+            return yearConstraint.nextValidValue(currentValue: currentValue)!
+        }
+
+    }
+
+    public func resolveNextDateThatSatisfiesRule(date: Date) throws -> Date {
```

Comment:
I think this `date: ` parameter could be more clear. Is it the current date that should be passed? Or just any arbitrary date?